scripts/refresh-sharding-and-restart

```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env bash
+
+set -e
+
+args="$(getopt -o '' -l adjusted-only,shards-added,shards-removed -n "$0" -- "$@")"
+eval "set -- $args"
+while true; do
+    case "$1" in
+        --adjusted-only) ADJUSTED_ONLY=1; shift;;
+        --shards-added) SHARDS_ADDED=1; shift;;
+        --shards-removed) SHARDS_REMOVED=1; shift;;
+        --) shift; break;;
+    esac
+done
+
+if [ -z "$ADJUSTED_ONLY" ] && [ -z "$SHARDS_ADDED" ] && [ -z "$SHARDS_REMOVED" ]; then
+    echo "You need to specify a valid option."
+    exit 1
+fi
+
+"$(dirname "$0")/zulip-puppet-apply" -f
+
+# In the ordering of operationss below the crucial detail
+# is that zulip-django and zulip-workers:* need to be restarted
+# before reloading nginx. Django has an in-memory map of which realm
+# belongs to which shard. Reloading nginx will cause users' tornado
+# requests to be routed according to the new sharding scheme. If that happens
+# before Django is restarted, updating its realm->shard map, users on realms,
+# whose shard has changed, will have their tornado requests handled by the new
+# tornado process, while Django will still use the old process for its internal
+# communication with tornado when servicing the user's requests.
+# That's a bad state that leads to clients getting into reload loops
+# ending in crashing on 500 response while Django is restarting.
+# For this reason it's important to reload nginx only after Django.
+if [ -n "$ADJUSTED_ONLY" ]; then
+    # The number of tornado workers hasn't changed, so we don't need
+    # to touch zulip-tornado:*.
+    supervisorctl restart zulip-django zulip-workers:*
+    service nginx reload
+elif [ -n "$SHARDS_ADDED" ]; then
+    # An additional tornado process is needed, so first
+    # we restart zulip-tornado:* to have the process ready to accept requests.
+    # TODO: This can be significantly improved, because we can simply start
+    # the new tornado process without restarting the existing ones.
+    supervisorctl restart zulip-tornado:*
+    supervisorctl restart zulip-django zulip-workers:*
+    service nginx reload
+elif [ -n "$SHARDS_REMOVED" ]; then
+    supervisorctl restart zulip-django zulip-workers:*
```

Comment:
We should always restart workers after doing the Django restart; there's a nonlinear supervisord performance cost to doing a lot of jobs at once, and zulip-django being down is a user-facing outage.